src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -212,9 +213,21 @@ public WebApplication Build()
 
             _builtApplication = new WebApplication(_hostBuilder.Build());
 
+            // We know that ConfigurationBuilder constructs ConfigurationRoot with a List<IConfigurationProvider> and all the
+            // IgnoreLoadConfigurationProviders have been "loaded", so we can put the originals back.
+            var providerList = (List<IConfigurationProvider>)((IConfigurationRoot)_builtApplication.Configuration).Providers;
+            for (var i = 0; i < providerList.Count; i++)
+            {
+                if (providerList[i] is IgnoreLoadConfigurationProvider wrappedProvider)
+                {
+                    providerList[i] = wrappedProvider.OriginalProvider;
+                }
+            }
+
             // Make builder.Configuration match the final configuration. To do that
-            // we clear the sources and add the built configuration as a source
-            ((IConfigurationBuilder)Configuration).Sources.Clear();
+            // we recreate the ConfigurationManager and add the built configuration as a source.
+            // We do not clear or dispose the old ConfigurationManager as this would dispose IConfigurationProviders that are in use.
+            Configuration = new ConfigurationManager();
```

Comment:
We need to fix the regression here where the captured reference doesn't see the final configuration.  I wonder if we can pull off a similar trick that we do with the  `IConfigurationBuilder`. If only the list of providers here wasn't a copy...

https://github.com/dotnet/runtime/blob/51e4954e785732cc551a841e9567d7bf2073ee21/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs#L75-L86

Ideally we wouldn't use private reflection here but it looks like the only answer currently...